cmd.go

```diff
@@ -146,23 +142,23 @@ func main() {
 			err = checkWritable(*fileArg)
 			checkError(err)
 		}
-		var r io.Reader
+		var r io.ReadCloser
 		if dataPipedIn() {
-			r = os.Stdin
+			r = ioutil.NopCloser(os.Stdin)
 		} else {
 			f, err := os.Open(*fileArg)
 			checkError(err)
-			defer f.Close()
 			r = f
 		}
 		if *dryRunArg {
 			err = doAdd(r, os.Stdout, addflags.Args())
 			checkError(err)
 		} else {
-			tmp, err := ioutil.TempFile("/tmp", "hostsfile-temp")
+			tmp, err := tempFile(*fileArg)
 			checkError(err)
-			defer tmp.Close()
 			err = doAdd(r, tmp, addflags.Args())
+			r.Close()
+			tmp.Close()
```

Comment:
I think we need to check the errors here, certainly an error closing a file that we have written to may indicate that we did not write the entire file, and it's probably a good practice to check the error on `r.Close()` as well.

Reply:
You're right. Although I don't think errors were checked in the original code either (`Close()` were deferred).

The reason I [changed the code](https://github.com/lende/hostsfile/commit/e7ff85bdc008fff5dd42d06097e45a7a004bdf7b) were to close the files as soon as possible, so Windows wouldn't complain about open files when renaming (and fail). I should have noted this somewhere...

Reply:
After a bit of research, it seems there is little point in checking for errors when closing a file that were opened read-only (see the official Go source for many examples).

Reply:
We write data to `tmp` though, right? So we need to check errors there.

Reply:
Yes